src/test/java/com/google/finapp/FinAppIT.java

```diff
@@ -124,4 +128,140 @@ public void createAccount_createsSingleValidAccount() throws Exception {
       assertThat(count).isEqualTo(1);
     }
   }
+
+  @Test
+  public void moveAccountBalance_validTransfer() throws Exception {
+    ByteArray fromAccountId = UuidConverter.getBytesFromUuid(UUID.randomUUID());
+    ByteArray toAccountId = UuidConverter.getBytesFromUuid(UUID.randomUUID());
+    BigDecimal fromAccountBalance = new BigDecimal(20);
+    BigDecimal toAccountBalance = new BigDecimal(0);
+    BigDecimal amount = new BigDecimal(10);
+    databaseClient.write(
+        ImmutableList.of(
+            Mutation.newInsertBuilder("Account")
+                .set("AccountId")
+                .to(fromAccountId)
+                .set("AccountType")
+                .to(AccountType.UNSPECIFIED_ACCOUNT_TYPE.getNumber())
+                .set("AccountStatus")
+                .to(AccountStatus.UNSPECIFIED_ACCOUNT_STATUS.getNumber())
+                .set("Balance")
+                .to(fromAccountBalance)
+                .set("CreationTimestamp")
+                .to(Value.COMMIT_TIMESTAMP)
+                .build(),
+            Mutation.newInsertBuilder("Account")
+                .set("AccountId")
+                .to(toAccountId)
+                .set("AccountType")
+                .to(AccountType.UNSPECIFIED_ACCOUNT_TYPE.getNumber())
+                .set("AccountStatus")
+                .to(AccountStatus.UNSPECIFIED_ACCOUNT_STATUS.getNumber())
+                .set("Balance")
+                .to(toAccountBalance)
+                .set("CreationTimestamp")
+                .to(Value.COMMIT_TIMESTAMP)
+                .build()));
+    spannerDao.moveAccountBalance(fromAccountId, toAccountId, amount);
+    try (ResultSet resultSet =
+        databaseClient
+            .singleUse()
+            .read(
+                "Account",
+                KeySet.newBuilder()
+                    .addKey(Key.of(fromAccountId))
+                    .addKey(Key.of(toAccountId))
+                    .build(),
+                Arrays.asList("AccountId", "Balance"))) {
+      int count = 0;
+      while (resultSet.next()) {
+        if (resultSet.getBytes(0).equals(fromAccountId)) {
+          assertThat(resultSet.getBigDecimal(1)).isEqualTo(fromAccountBalance.subtract(amount));
+          count++;
+        } else if (resultSet.getBytes(0).equals(toAccountId)) {
+          assertThat(resultSet.getBigDecimal(1)).isEqualTo(toAccountBalance.add(amount));
+          count++;
+        }
+      }
+      assertThat(count).isEqualTo(2);
+    }
+  }
+
+  @Test
+  public void moveAccountBalance_negativeAmount_throwsException() throws Exception {
```

Comment:
We don't have throwing `IllegalArgumentException`s in the signature, so it's debatable of whether we should even test this or not. Or if we should add this to either the signature and/or the docstring

Reply:
What about using a more specific custom checked exception instead? Then it's in the signature and docs, and gives us a more specific error.

It's possible that `SpannerDaoImpl.moveAccountBalance` could throw an `IllegalArgumentException` that isn't because of a bad argument from the end user, but because of some bad logic handling the request. But if the request would make the balance go negative that's clearly a bad argument from the user and we should send back the right gRPC status code (3).

Reply:
https://github.com/GoogleCloudPlatform/cloud-spanner-samples/issues/34